### PR TITLE
HDDS-8995. Refactor the datanode BlockDeletingService

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

This pr refactors the BlockDeletingService.

At present, DeleteBlockService in datanode is bound to KeyValueContainer, in this PR, BlockDeletingService is abstracted to common, and DeleteBlockTask is implemented in keyvalue container, which realizes decoupling, more conducive to the subsequent expansion of other types of containers on datanode.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8995

## How was this patch tested?

ut

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
PTAL @ChenSammi Thanks.

**Comment by guohao-rosicky** [comment]
ci passed, see: [CI](https://github.com/guohao-rosicky/ozone/actions/runs/5527333803)

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/ContainerUtils.java [hunk 1]
```diff
@@ -326,4 +327,15 @@ public static long retrieveContainerIdFromTarName(String tarName)
...
   }
+
+  public static long getPendingDeletionBlocks(ContainerData containerData) {
+    if (containerData.getContainerType()
+        .equals(ContainerProtos.ContainerType.KeyValueContainer)) {
+      return ((KeyValueContainerData) containerData)
+          .getNumPendingDeletionBlocks();
+    } else {
+      // If another ContainerType is available later, implement it
+      return 0L;
```
Comment (ChenSammi) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/ContainerUtils.java:338]: Please throw the exception here.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java [hunk 2]
```diff
@@ -546,4 +546,8 @@ StorageVolumeChecker getVolumeChecker(ConfigurationSource conf) {
   public ContainerMetrics getMetrics() {
     return metrics;
   }
+
+  public BlockDeletingService getBlockDeletingService() {
```
Comment (ChenSammi) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java:550]: This API is not used anywhere. Suggest to remove it.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/BlockDeletingService.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+      long totalBlocks = 0;
+      for (ContainerBlockInfo containerBlockInfo : containers) {
+        BlockDeletingTaskBuilder builder =
+            new BlockDeletingTaskBuilder();
+        builder.setBlockDeletingService(this)
+            .setContainerBlockInfo(containerBlockInfo)
+            .setPriority(TASK_PRIORITY_DEFAULT);
+        containerBlockInfos = builder.build();
+        queue.add(containerBlockInfos);
+        totalBlocks += containerBlockInfo.getBlocks();
```
Comment (ChenSammi) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/BlockDeletingService.java:133]: the deleted blocks count is containerBlockInfo.numBlocksToDelete.

**Comment by ChenSammi** [comment]
The last patch LGTM.  Wait for the CI to pass.

**Comment by guohao-rosicky** [comment]
ci passed, see: https://github.com/guohao-rosicky/ozone/actions/runs/5959164796

**Comment by ChenSammi** [comment]
Thanks @guohao-rosicky for the contribution.

**Review by ChenSammi — APPROVED** [review]

**Pull request merged**